{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dd10f621",
   "metadata": {},
   "source": [
    "# NCBI Databases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "048ca987",
   "metadata": {},
   "source": [
    "Name at least five different databases of the NCBI Entrez system. Write a short description for each named database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "403c39d3",
   "metadata": {},
   "source": [
    "1. PubMed database: contains millions of citations from life science journals. A great deal of these citations have abstracts, and millions have links to their full text articles (or both an abstract and a link to full text. PubMed is heavily linked to other core NCBI databases, thereby providing a crucial bridge between the data of molecular biology and the scientific\n",
    "literature.\n",
    "2. Taxonomy database: provides links to all data for each taxonomic node, from superkingdoms to subspecies. The taxonomy database reflects sequence data from a great deal of formally\n",
    "described species. The Taxonomy Browser can be used to view the taxonomy tree or retrieve data from any of the Entrez databases for a particular organism or group.\n",
    "3. RefSeq (Reference Sequence) database: allows a user access to non-redundant set of curated and computationally derived sequences for transcripts, proteins and genomic regions. RefSeq DNA and RNA sequences can be searched and retrieved from the Nucleotide database.\n",
    "4. SRA (Sequence Read Archive) database: contains data generated by the latest generation of high-throughput nucleic acid sequencers.SRA stores raw sequence reads and alignments. Data are deposited into SRA as supporting evidence for a wide range of study types including de novo genome assemblies, GWAS, single nucleotide polymorphism and structural variation analysis, pathogen identification, transcript assembly, metagenomic community profiling and epigenetics.\n",
    "5. Epigenomics database: collects data from studies examining epigenetic features such as post-translational modifications of histone proteins, genomic DNA methylation, chromatin organization and the expression of non-coding regulatory RNA. Raw data from these\n",
    "experiments, together with extensive metadata, are stored in the GEO and SRA databases. Data arepremapped to genomic coordinates, so users are not required to be familiar with or\n",
    "manipulate the raw data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe8376f",
   "metadata": {},
   "source": [
    "# Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b5f6b2e",
   "metadata": {},
   "source": [
    "Write down the example syntax for the EInfo, ESearch, ESummary, EFetch and ELink method as you did during your exercises. You can run all methods in one script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "716d0a4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.79'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import Bio\n",
    "Bio.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "12ca8e90",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez as ez\n",
    "ez.email = 'biopython@gmail.com' # fake email address"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "207b2fda",
   "metadata": {},
   "source": [
    "## EInfo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1b5eb8f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Entrez.Parser.DictionaryElement'>\n",
      "dict_keys(['DbList'])\n",
      "['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'ncbisearch', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'biosystems', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']\n"
     ]
    }
   ],
   "source": [
    "# EInfo (to get a list of DBs)\n",
    "record_1 = ez.read(ez.einfo())\n",
    "print(type(record_1))\n",
    "print(record_1.keys())\n",
    "print(record_1['DbList'])\n",
    "# print('\\n'.join(sorted(record_1['DbList']))) # to make a string with each db at a new line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "b2040b05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Entrez.Parser.DictionaryElement'>\n",
      "dict_keys(['DbInfo'])\n",
      "DbName : genome\n",
      "MenuName : Genome\n",
      "Description : Genomic sequences, contigs, and maps\n",
      "DbBuild : Build220322-0605.1\n",
      "Count : 86051\n",
      "LastUpdate : 2022/03/22 09:36\n",
      "LinkList : [{'Name': 'genome_assembly', 'Menu': 'Assembly', 'Description': 'Related Assembly records', 'DbTo': 'assembly'}, {'Name': 'genome_bioproject', 'Menu': 'BioProject Links', 'Description': 'Related BioProjects', 'DbTo': 'bioproject'}, {'Name': 'genome_gene', 'Menu': 'Gene Links', 'Description': 'Related Gene', 'DbTo': 'gene'}, {'Name': 'genome_nuccore', 'Menu': 'Components', 'Description': 'Constituent nucleotide sequence', 'DbTo': 'nuccore'}, {'Name': 'genome_nuccore_samespecies', 'Menu': 'Other genomes for species', 'Description': 'Other (nearly) complete genomes or particular genomic segments for the species from DDBJ/EMBL/GenBank', 'DbTo': 'nuccore'}, {'Name': 'genome_protein', 'Menu': 'Protein Links', 'Description': 'Constituent protein sequence', 'DbTo': 'protein'}, {'Name': 'genome_proteinclusters', 'Menu': 'Protein Cluster Links', 'Description': 'Related Protein Clusters', 'DbTo': 'proteinclusters'}, {'Name': 'genome_pubmed', 'Menu': 'PubMed Links', 'Description': 'PubMed articles cited by genome record', 'DbTo': 'pubmed'}, {'Name': 'genome_taxonomy', 'Menu': 'Taxonomy Links', 'Description': 'Taxonomy sequences associated with genome record', 'DbTo': 'taxonomy'}]\n"
     ]
    }
   ],
   "source": [
    "# EInfo (e.g. for the Genome database)\n",
    "record_2 = ez.read(ez.einfo(db = 'genome'))\n",
    "print(type(record_2))\n",
    "print(record_2.keys())\n",
    "#print(record_2['DbInfo'])\n",
    "for key in record_2['DbInfo'].keys():\n",
    "    if key != 'FieldList':\n",
    "        print(key, ':', record_2['DbInfo'][key])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6697bfb",
   "metadata": {},
   "source": [
    "## ESearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "ac7d93f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Entrez.Parser.DictionaryElement'>\n",
      "dict_keys(['Count', 'RetMax', 'RetStart', 'IdList', 'TranslationSet', 'TranslationStack', 'QueryTranslation'])\n",
      "Count : 7\n",
      "RetMax : 7\n",
      "RetStart : 0\n",
      "IdList : ['24132078', '15132172', '21156594', '24221179', '7242651', '4939438', '13490378']\n",
      "TranslationSet : []\n",
      "TranslationStack : [{'Term': 'Badhamia[All Fields]', 'Field': 'All Fields', 'Count': '7', 'Explode': 'N'}, 'GROUP']\n",
      "QueryTranslation : Badhamia[All Fields]\n",
      "['24132078', '15132172', '21156594', '24221179', '7242651', '4939438', '13490378']\n"
     ]
    }
   ],
   "source": [
    "# ESearch\n",
    "'''Comment: I googled and found out that Badhamia lilacina is not very well studied. So, I decided\n",
    "to do my search against Badhamia (term 'Badhamia lilacina' gives only one count'''\n",
    "record_3 = ez.read(ez.esearch(db ='pubmed', term = 'Badhamia', retmax = 7))\n",
    "print(type(record_3))\n",
    "print(record_3.keys())\n",
    "for key in record_3.keys():\n",
    "    print(key, ':', record_3[key])\n",
    "BadhamiaIDs = record_3['IdList']\n",
    "print(BadhamiaIDs) # list of PubMed IDs\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cf345b5",
   "metadata": {},
   "source": [
    "## ESummary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "9265fa3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Date of publication:\n",
      "2014 Apr\n",
      "First author:\n",
      "Aguilar M\n",
      "Title:\n",
      "Using environmental niche models to test the 'everything is everywhere' hypothesis for Badhamia.\n",
      "\n",
      "Date of publication:\n",
      "2003 Jul-Aug\n",
      "First author:\n",
      "Haugen P\n",
      "Title:\n",
      "The molecular evolution and structural organization of self-splicing group I introns at position 516 in nuclear SSU rDNA of myxomycetes.\n",
      "\n",
      "Date of publication:\n",
      "2003 Jan-Feb\n",
      "First author:\n",
      "Clark J\n",
      "Title:\n",
      "Biosystematics of the myxomycete Badhamia gracilis.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ESummary\n",
    "'''Here I use my list of IDs I created searching against \"Badhamia\", but I limit it up to 3 records\n",
    "only to descrease in size my potential output'''\n",
    "for ID in BadhamiaIDs[:3]:\n",
    "    jr_summary = ez.read(ez.esummary(db = 'pubmed', id=ID))\n",
    "    #print(jr_summary[0].keys())\n",
    "    for handle in jr_summary:\n",
    "        print(f\"Date of publication:\\n{handle['PubDate']}\\nFirst author:\\n{handle['AuthorList'][0]}\\nTitle:\\n{handle['Title']}\\n\")\n",
    "\n",
    "# Here I am prinditng out dates of publications, the first authors for each paper and the titles."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59fccf4a",
   "metadata": {},
   "source": [
    "## EFetch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "6afaf53a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23\n",
      ">NM_003140.3 Homo sapiens sex determining region Y (SRY), mRNA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# EFetch\n",
    "# https://medlineplus.gov/genetics/gene/sry/\n",
    "'''Just for fun decided to play around with SRY gene - a gender-determining one'''\n",
    "# SRY: sex-determining region Y protein\n",
    "record_4 = ez.read(ez.esearch(db = 'nucleotide', term = 'SRY[Gene Name] AND RefSeq[Keyword]', retmax=2000, idtype = 'acc'))\n",
    "#print(record_4)\n",
    "counter_1 = 0\n",
    "fetchList = []\n",
    "for ID in record_4['IdList']:\n",
    "    if 'NM_' in ID: # catches only mRNA\n",
    "        counter_1 += 1\n",
    "        fetch = ez.efetch(db = 'nucleotide', id = ID, rettype='fasta', retmode = 'text')\n",
    "        #readFetch = fetch.readline()\n",
    "        readFetch = fetch.read()\n",
    "        fetchList.append(readFetch)\n",
    "        #print(readFetch)\n",
    "#print(fetchList)\n",
    "print(len(fetchList))\n",
    "#print(f\"Number of the NM_ records found: {counter_1}\") # outputs only 23 records\n",
    "\n",
    "for files in fetchList:\n",
    "    with open('SRY.fasta', 'a+') as savedSRY:\n",
    "        savedSRY.write(files) # now there is a SRY.fasta file in my current directory\n",
    "with open ('SRY.fasta', 'r') as f:\n",
    "    first_line = next(f)\n",
    "# printing out only the first line of the long fasta file to check that the code is working all right\n",
    "print(first_line) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b8968d2",
   "metadata": {},
   "source": [
    "## ELink"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a1407bbd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['24132078', '15132172', '21156594', '24221179', '7242651', '4939438', '13490378']\n",
      "24132078\n",
      "dict_keys(['ERROR', 'LinkSetDb', 'LinkSetDbHistory', 'DbFrom', 'IdList'])\n",
      "Database: pubmed, ID: ['24132078']\n",
      "Number of crosslinks found: 7\n",
      "Number of PubMed IDs found: 105\n",
      "['24132078', '28414791', '22402402', '25550302', '32649270', '22530009', '26663217', '18522694', '19750951', '29533140', '25028491', '25232071', '27128786', '26702956', '31586669', '23667494', '29024429', '29953506', '20456231', '31613722', '31059141', '22962350', '17854485', '23497060', '26321302', '33311142', '20524603', '34676941', '18836695', '24486990', '19197466', '33171353', '31874965', '28919504', '29167159', '15132172', '30903202', '22492409', '34913576', '24433389', '22818199', '28647618', '33966339', '25451805', '27190206', '31055860', '17045493', '26240305', '21262984', '31582156', '29337274', '32829066', '29248627', '26432394', '17062415', '33090298', '26095958', '16367842', '33187474', '22776548', '25300454', '17603517', '24018859', '23998707', '31358042', '27593700', '24118115', '18761283', '22970129', '24987129', '17306538', '25732069', '34624792', '30187122', '22421085', '32405644', '23869920', '27351595', '31904883', '19168565', '22451600', '23167902', '27233443', '32399960', '15212388', '30076328', '10633916', '22319123', '20943518', '22516430', '30783861', '27539600', '17594425', '30945986', '28487259', '28187628', '22112441', '26336224', '20594354', '17560828', '20525639', '20205541', '22906180', '24838118', '34533412']\n",
      "E.g. this paper with ID 28414791 has in references my target paper, ref. 27\n"
     ]
    }
   ],
   "source": [
    "# ELink\n",
    "'''I will use my previously created list of IDs for Badhamia'''\n",
    "print(BadhamiaIDs)\n",
    "pmid = BadhamiaIDs[0] # get the first paper in my list PMID 24132078\n",
    "print(pmid) # PMCID is 3960529)\n",
    "record_5 = ez.read(ez.elink(dbfrom=\"pubmed\", id=pmid))\n",
    "#print(type(record_5)) # list\n",
    "#print(record_5[0]) # getting into dictionary\n",
    "print(record_5[0].keys())\n",
    "print(f\"Database: {record_5[0]['DbFrom']}, ID: {record_5[0]['IdList']}\")\n",
    "# LinkSetDb - contains my search results\n",
    "print(f'Number of crosslinks found: {len(record_5[0][\"LinkSetDb\"])}') # I found 7 crosslinks! \n",
    "# print(type(record_5[0][\"LinkSetDb\"])) # list\n",
    "#print(record_5[0][\"LinkSetDb\"][0]) # dictionary\n",
    "#print(record_5[0][\"LinkSetDb\"][0]['Link']) # list of dictionaries\n",
    "counter_2 = 0\n",
    "pmIDlist = []\n",
    "for link in record_5[0][\"LinkSetDb\"][0]['Link']:\n",
    "    #print(link['Id'])\n",
    "    counter_2 += 1\n",
    "    pmIDlist.append(link['Id'])\n",
    "print(f'Number of PubMed IDs found: {counter_2}') # found 105 PubMed crosslinked IDs.\n",
    "print(pmIDlist)\n",
    "print(f'E.g. this paper with ID {pmIDlist[1]} has in references my target paper, ref. 27') # e.g this paper has a reference to my target paper ref. 27"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Необработанный формат ячейки",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
